### PR TITLE
Add support for Cygwin repo

### LINKED ISSUE #269: Cygwin support
@tkelman:
> Cygwin: http://mirrors.mit.edu/cygwin/x86_64/setup.ini, http://mirrors.mit.edu/cygwin/x86/setup.ini

### PR DESCRIPTION
**Pull request by jon-turney**
Closes #269

### CODE REVIEW

File: repology/parsers/parsers/cygwin.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                    continue
+
+                pkg = factory.begin()
+
+                pkg.set_basename(package['name'])
+                pkg.add_categories([c.lower() for c in package['categories']])
+                pkg.set_summary(package['summary'])
+
+                if 'maintainers' in package:
+                    pkg.add_maintainers([m.lower().replace(' ', '.') + '@cygwin' for m in package['maintainers']])
```
[the remark this answers is not in the document]
Comment (jon-turney) [reply repology/parsers/parsers/cygwin.py]: Well spotted

**Comment by jon-turney** [comment]
> Some minor changes suggested,

Updated to incorporate those suggested changes

> also a question: I see `.cygport` files contain `HOMEPAGE` and `SRC_URI`, couldn't these be propagated to the dump? 

Unfortunately there's an intermediate step which doesn't propagate those pieces of information, currently.  

> Also isn't there a way to have/generate a link to source `.cygport` for each package?

Unfortunately, for historical reasons, there's no canonical repository for our cygport files, currently.

Though both things would be good to add, certainly :)

File: repology/parsers/parsers/cygwin.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                for maturity in ['stable', 'test']:
+                    if maturity not in packagedata['versions']:
+                        continue
+
+                    version = packagedata['versions'][maturity][-1]
+
+                    for subpackage in packagedata['subpackages']:
+                        pkg = factory.begin()
+
+                        pkg.set_basename(packagedata['name'])
```
Comment (AMDmi3) [inline repology/parsers/parsers/cygwin.py]: Sorry, perhaps I wasn't too clear. I haven't meant generic fields like basename, categories, summary and maintainers to be set in the inner loop. They belong outside of the loop, and this makes logs produced by PackageMakager more useful as basename can be used to identify the package which produces output. In the inner loop, just `clone` pkg and set subpackage specific fields like version, flags and name.

File: repos.d/cygwin.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+###########################################################################
+# Cygwin
+###########################################################################
+- name: cygwin
+  type: repository
+  desc: Cygwin
+  family: cygwin
+  color: '00fb15'
+  minpackages: 7000
```
Comment (AMDmi3) [inline repos.d/cygwin.yaml]: There are packages with no maintainers defined (at least `alternatives`). You may want to set e.g. `default_maintainer: orphaned@cygwin` here to be able to list these in repology.

File: repology/parsers/parsers/cygwin.py (hunk 2 above)
Comment (jon-turney) [reply repology/parsers/parsers/cygwin.py]: Ah, right.  I think I understand what you meant now.

**Comment by jon-turney** [comment]
Moved the check for '_obsolete' category from package to subpackage.  Among the effects this was having was suppressing a lot of python packages.

File: repology/parsers/parsers/cygwin.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                    if maturity not in package['versions']:
+                        continue
+
+                    version = package['versions'][maturity][-1]
+                    pkg.set_version(version.split('-', 1)[0])
+                    pkg.set_rawversion(version)
+                    pkg.set_flags(PackageFlags.devel, maturity == 'test')
+
+                    for subpackage in package['subpackages']:
+                        if subpackage['name'].endswith('-debuginfo'):
```
[the remark this answers is not in the document]
Comment (jon-turney) [reply repology/parsers/parsers/cygwin.py]: I dropped this, so debuginfo packages are now included.  But I'm not sure how this is supposed to interact with classification (i.e. dragon-debuginfo is not part of dragon-player, but dragon5-part is, somehow, which is confusing to me...)

**Comment by jon-turney** [comment]
Use `rsplit('-', 1)` rather than `split('-', 1)` to split V-R, to handle correctly some antique packages which have a hyphen in the version.

**Pull request merged**